### PR TITLE
Adds the Engineer Dispenser

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
Adds a Dispenser for the engineer, currently it heals nearby humans and allows them to carry alot of items in, might make it do other stuff on future, its equipped on backpack slot
image

## Why It's Good For The Game
Lets engineers drag around alot of items with them without being immediately accessible and also gives them a small healing utility(similar to tricordrazine but only burn and brute).

## Changelog
:cl:
add: Added the engineer dispenser which is equipped on back slot and carries alot of items, can only heal and access items when deployed. 
imageadd: new title screen announcing the new engineer dispenser
/:cl:

### CODE REVIEW

**Comment by ilikesaying** [comment]
https://www.youtube.com/watch?v=SNgNBsCI4EA

File: code/__DEFINES/equipment.dm [hunk 1]
```diff
@@ -67,9 +67,10 @@
 #define IS_DEPLOYABLE (1<<13) //Item can be deployed into a machine
 #define DEPLOY_ON_INITIALIZE (1<<14)
 #define IS_DEPLOYED (1<<15) //If this is on an item, said item is currently deployed
-#define DEPLOYED_NO_PICKUP  (1<<16) //Disables deployed item pickup
-#define DEPLOYED_NO_ROTATE  (1<<17) //Disables deployed item rotation abilities to rotate.
-#define DEPLOYED_WRENCH_DISASSEMBLE (1<<18) //If this is on an item, the item can only be disassembled using a wrench once deployed.
+	#define IS_DEPLOYING (1<<16) //If this is on an item, said item is currently deploying
```
Comment (BraveMole) [inline code/__DEFINES/equipment.dm:70]: probably should be a trait?

File: code/game/objects/items/storage/backpack.dm [hunk 2]
```diff
@@ -777,3 +778,131 @@
...
+/obj/item/storage/backpack/dispenser
+	name = "Dispense-O-Matic 9000 Provisions Dispenser"
+	desc = "The Dispense-O-Matic also known as \"Dispenser\" is a machine capable of holding a big amount of items on it, while also healing nearby humans. Your allies will often ask you to lay down one of those."
+	icon = 'icons/obj/items/storage/storage_48.dmi'
+	icon_state = "dispenser"
+	flags_equip_slot = ITEM_SLOT_BACK
+	max_w_class = 6
+	max_storage_space = 63
+	COOLDOWN_DECLARE(deploy_cooldown)
+	var/list/affecting_list
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:791]: doc

File: code/game/objects/items/storage/backpack.dm [hunk 3]
```diff
@@ -777,3 +778,131 @@
...
+
+/obj/item/storage/backpack/dispenser/open(mob/user)
+	if(loc == user)
+		return FALSE
+	. = ..()
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:796]: return ..()

File: code/game/objects/items/storage/backpack.dm [hunk 4]
```diff
@@ -777,3 +778,131 @@
...
+		open(user)
+		return
+	return ..()
+
+/obj/item/storage/backpack/dispenser/MouseDrop(obj/over_object)
+	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED))
+		if(over_object == usr && ishuman(over_object))
+			open(over_object)
+		return
+	. = ..()
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:810]: ditto

File: code/game/objects/items/storage/backpack.dm [hunk 5]
```diff
@@ -777,3 +778,131 @@
...
+
+/obj/item/storage/backpack/dispenser/attack_self(mob/user)
+	if(!ishuman(user) || CHECK_BITFIELD(flags_item, NODROP))
+		return
+	var/deploy_location
+	deploy_location = get_step(user, user.dir)
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:816]: one line declaration + say it's a turf

File: code/game/objects/items/storage/backpack.dm [hunk 6]
```diff
@@ -777,3 +778,131 @@
...
+	if(!do_after(user, 0.5 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		return
+
+	user.temporarilyRemoveItemFromInventory(src)
+
+	forceMove(deploy_location)
+	density = TRUE
+	anchored = TRUE
+	dir = REVERSE_DIR(user.dir)
+	icon_state = "dispenser_deployed"
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:833]: use update_icon_state

File: code/game/objects/items/storage/backpack.dm [hunk 7]
```diff
@@ -777,3 +778,131 @@
...
+	balloon_alert_to_viewers("Deploying...")
+	flick("dispenser_deploy", src)
+	ENABLE_BITFIELD(flags_item, IS_DEPLOYING)
+
+	addtimer(CALLBACK(src, .proc/deploy), 4.2 SECONDS)
+
+/obj/item/storage/backpack/dispenser/proc/deploy()
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:840]: doc

File: code/game/objects/items/storage/backpack.dm [hunk 8]
```diff
@@ -777,3 +778,131 @@
...
+	DISABLE_BITFIELD(flags_item, IS_DEPLOYING)
+	ENABLE_BITFIELD(flags_item, IS_DEPLOYED)
+	affecting_list = list()
+	for(var/mob/living/carbon/human/human in view(3))
+		if(!line_of_sight(src, human, 2))
+			continue
+		RegisterSignal(human, COMSIG_PARENT_QDELETING, .proc/on_affecting_qdel)
+		affecting_list[human] = beam(human, "blood_light")
+		human.playsound_local(get_turf(src), 'sound/machines/dispenser/dispenser_heal.ogg', 50)
+	for(var/turf/turfs in range(2, src))
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:852]: AS

Comment (BraveMole) [reply code/game/objects/items/storage/backpack.dm:852]: and turf with no s, or T

Comment (BraveMole) [reply code/game/objects/items/storage/backpack.dm:852]: and turf with no s, or T

File: code/game/objects/items/storage/backpack.dm [hunk 9]
```diff
@@ -777,3 +778,131 @@
...
+/obj/item/storage/backpack/dispenser/proc/on_affecting_qdel(datum/source)
+	affecting_list -= source
+
+/obj/item/storage/backpack/dispenser/CtrlClick(mob/user)
+	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED) && !CHECK_BITFIELD(flags_item, IS_DEPLOYING))
+		if(!can_interact(user))
+			return
+		balloon_alert_to_viewers("Undeploying...")
+
+		icon_state = "dispenser"
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:865]: update_icon_state

File: code/game/objects/items/storage/backpack.dm [hunk 10]
```diff
@@ -777,3 +778,131 @@
...
+		flick("dispenser_undeploy", src)
+		ENABLE_BITFIELD(flags_item, IS_DEPLOYING)
+		for(var/turf/turfs in range(2, src))
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:868]: turf, AS in

File: code/game/objects/items/storage/backpack.dm [hunk 11]
```diff
@@ -777,3 +778,131 @@
...
+			qdel(affecting_list[affecting])
+			UnregisterSignal(affecting, COMSIG_PARENT_QDELETING)
+		affecting_list = null
+		STOP_PROCESSING(SSobj, src)
+		addtimer(CALLBACK(src, .proc/undeploy), 4.2 SECONDS)
+		return
+	return ..()
+
+/obj/item/storage/backpack/dispenser/proc/undeploy()
+	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED))
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:880]: early return, autodoc

File: code/game/objects/items/storage/backpack.dm [hunk 12]
```diff
@@ -777,3 +778,131 @@
...
+		density = FALSE
+		anchored = FALSE
+		DISABLE_BITFIELD(flags_item, IS_DEPLOYED)
+		DISABLE_BITFIELD(flags_item, IS_DEPLOYING)
+
+
+
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:887]: random white lines

File: code/game/objects/items/storage/backpack.dm [hunk 13]
```diff
@@ -777,3 +778,131 @@
...
+/obj/item/storage/backpack/dispenser/process()
+	for(var/mob/living/carbon/human/affecting AS in affecting_list)
+		if(!line_of_sight(src, affecting, 3))
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:890]: 3 should be a range define

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Comment by Walarks** [comment]
*It has been 90 years...*

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Please find a use that isnt healing, medical is already in need of untangling without free healing like this

**Comment by EvilMelonTheSecond** [comment]
Should probably make it dispense free gauze and ointment instead

**Comment by Luxyboo** [comment]
maybe let it passively generate metal every 5 minutes

**Comment by BraveMole** [comment]
image

**Pull request closed without merging**